client/state/analytics/README.md

```diff
@@ -0,0 +1,90 @@
+# Analytics Middleware
+
+## Overview
+
+Analytics track things. By their very nature they are side-effecting actions, usually by sending API calls to one or more analytics servers.
+ 
+Calypso supports a variety of analytics methods including (but not limited to) Google Analytics, stat bumping, and Tracks events.
+ 
+As we develop the different view components in React for Calypso, we should work hard to keep them as pure and simple as they can be. This brings with it many benefits, including the ability to easily test those components. Calling out these side-effecting operations can break those tests among other things.
+ 
+_**Middleware**_ is a stage in the **Redux** pipeline that runs between dispatching an action and when that action runs through the reducers. By incorporating the side-effecting operations at this point, we decouple those breaking effects from the other parts of the app.
+ 
+This module provides just such a middleware: it allows React components to now fire off Redux actions indicating the _intent_ to send an analytics metric instead of actually sending it, keeping the otherwise-unreleted dependencies out of the components.
+ 
+## Usage
+
+As this middleware follows the [Flux Standard Action](https://github.com/acdlite/flux-standard-action) pattern, you can start tracking analytics simply by adding a new `meta` block to your existing actions. A helper function `withAnalytics()` is provided to make this easy.
+
+A suite of pure analytics action creators are also exposed which can be used in combination with `withAnalytics()` or by themselves if no other action beyond analytics tracking is required.
+
+### Examples
+
+```js
+import {
+    withAnalytics,
+    bumpStat,
+    recordGoogleEvent,
+    recordTracksEvent,
+    recordPageView
+} from 'state/analytics/actions';
+
+// track a page-view
+dispatch( recordPageView( '/path/to/page', 'Page Title' ) )
+
+// add event-tracking to an action
+dispatch( withAnalytics(
+    recordGoogleEvent( 'selected_thing', 'my_thing' ),
+    selectThing( myThing )
+) );
+
+// withAnalytics() is auto-curried for convenience
+const statBumper = withAnalytics( bumpStat( 'api_calls', 'success' ) );
+dispatch( statBumper( apiSuccessAction() ) );
+
+
+// works with pure actions and thunks
+dispatch( withAnalytics(
+    recordPageView( '/api/users', 'Fetch Users' ),
+    fetchUsers( siteId ) // returns a thunk which makes an API call
+) );
+
+// passed as a component prop
+const trackSelection = page =>
+    withAnalytics( recordTrackEvent( 'selected_page', { 'page': page.name } ) );
+    
+const mapDispatchToProps = dispatch => ( {
+    selectPage: page => () => dispatch( trackSelection( page )( setCurrentPage( page ) ) ),
+    recordPageLoad: () => dispatch( bumpStat( 'page_loaded', 'page_selected_page' ) )
+} );
```

Comment:
I could change it, but in the example I have chosen I would still need to provide an object because of the way `selectPage` is written. Correspondingly, I could change the example to be simpler.

My thoughts were that we're all free to add helpers later but I wanted to provide a minimal example.

Very flexible on this point :wink: